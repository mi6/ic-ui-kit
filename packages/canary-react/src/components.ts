/* eslint-disable */
/* tslint:disable */
/* auto-generated react proxies */
import { createReactComponent } from './react-component-lib';

import type { JSX } from '@ukic/canary-web-components';

import { defineCustomElements } from '@ukic/canary-web-components/loader';

defineCustomElements();
export const IcDataTable = /*@__PURE__*/createReactComponent<JSX.IcDataTable, HTMLIcDataTableElement>('ic-data-table');
export const IcDateInput = /*@__PURE__*/createReactComponent<JSX.IcDateInput, HTMLIcDateInputElement>('ic-date-input');
export const IcMenuWithMulti = /*@__PURE__*/createReactComponent<JSX.IcMenuWithMulti, HTMLIcMenuWithMultiElement>('ic-menu-with-multi');
export const IcPaginationBar = /*@__PURE__*/createReactComponent<JSX.IcPaginationBar, HTMLIcPaginationBarElement>('ic-pagination-bar');
export const IcSelectWithMulti = /*@__PURE__*/createReactComponent<JSX.IcSelectWithMulti, HTMLIcSelectWithMultiElement>('ic-select-with-multi');
