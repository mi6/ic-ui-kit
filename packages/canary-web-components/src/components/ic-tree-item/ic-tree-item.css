:host {
  display: block;
  border: none;
  position: relative;
}

:host .tree-item-content {
  display: flex;
  align-items: center;
  height: calc(var(--ic-space-xl) + var(--ic-space-xs));
  padding: 0 var(--ic-space-xs);
  text-decoration: none;
  color: var(--ic-primary-text);

  --ic-typography-color: var(--ic-primary-text);
}

::slotted([slot="router-item"]) {
  display: flex;
  align-items: center;
  height: calc(var(--ic-space-xl) + var(--ic-space-xs));
  text-decoration: none;
  font-family: var(--ic-font-body-family);
  color: var(--ic-primary-text);
  padding: 0 var(--ic-space-xs) !important;
}

::slotted([slot="router-item"].ic-tree-item-single) {
  padding-left: calc(var(--ic-space-xl) + var(--ic-space-xs)) !important;
}

:host .tree-item-content:focus,
::slotted([slot="router-item"]:focus) {
  box-shadow: var(--ic-border-focus);
  border-radius: var(--ic-border-radius);
  transition: var(--ic-transition-duration-fast);
  outline: none;
}

:host .tree-item-content:hover,
::slotted([slot="router-item"]:hover) {
  background-color: var(--ic-action-dark-bg-hover) !important;
  cursor: pointer;
}

:host .tree-item-content:active,
::slotted([slot="router-item"]:active) {
  background-color: var(--ic-action-dark-bg-active) !important;
}

:host(.ic-tree-item-selected) .tree-item-content,
:host(.ic-tree-item-selected) ::slotted([slot="router-item"]) {
  background-color: var(--ic-action-default-bg-hover) !important;
}

:host(.ic-tree-item-small) .tree-item-content,
:host(.ic-tree-item-small) ::slotted([slot="router-item"]) {
  height: var(--ic-space-xl);
}

:host(.ic-tree-item-large) .tree-item-content,
:host(.ic-tree-item-large) ::slotted([slot="router-item"]) {
  height: var(--ic-space-xxl);
}

:host(.ic-tree-item-disabled) {
  pointer-events: none;
}

:host(.ic-tree-item-disabled) .tree-item-content,
:host(.ic-tree-item-disabled) .arrow-dropdown,
:host(.ic-tree-item-disabled) ::slotted([slot="icon"]),
:host(.ic-tree-item-disabled) ::slotted([slot="router-item"]) {
  color: var(--ic-architectural-200) !important;

  --ic-typography-color: var(--ic-architectural-200) !important;

  fill: var(--ic-architectural-200);
  pointer-events: none;
}

:host(.ic-tree-item-light) .tree-item-content,
:host(.ic-tree-item-light) .arrow-dropdown,
:host(.ic-tree-item-light) ::slotted([slot="icon"]),
:host(.ic-tree-item-light) ::slotted([slot="router-item"]) {
  color: var(--ic-architectural-white);

  --ic-typography-color: var(--ic-architectural-white);

  fill: var(--ic-architectural-white);
}

:host .arrow-dropdown {
  color: var(--ic-action-default);
  margin-right: var(--ic-space-xs);
}

:host(.ic-tree-item-small) .arrow-dropdown {
  margin-right: var(--ic-space-xxs);
}

:host .tree-item-expanded {
  transform: rotate(180deg);
  margin-top: calc(-1 * var(--ic-space-xxs));
}

:host .ic-tree-item-single {
  padding-left: calc(var(--ic-space-xl) + var(--ic-space-xs));
}

:host(.ic-tree-item-small) .ic-tree-item-single {
  padding-left: calc(var(--ic-space-xl) + var(--ic-space-xxs));
}

.icon-container {
  width: var(--ic-space-lg);
  height: var(--ic-space-lg);
  margin: 0 var(--ic-space-xs) 0 0;
}

:host(.ic-tree-item-small) .icon-container {
  margin: 0 var(--ic-space-xxs) 0 0;
}

.ic-text-overflow {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.ic-tooltip-overflow {
  overflow: hidden;
}

/** High Contrast **/
@media (forced-colors: active) {
  ::slotted([slot="icon"]) {
    fill: currentcolor;
  }

  :host .tree-item-content:hover,
  ::slotted([slot="router-item"]:hover),
  :host .tree-item-content:active,
  ::slotted([slot="router-item"]:active),
  :host(.ic-tree-item-selected) .tree-item-content,
  :host(.ic-tree-item-selected) ::slotted([slot="router-item"]) {
    background-color: none !important;
  }

  :host(.ic-tree-item-disabled) .tree-item-content,
  :host(.ic-tree-item-disabled) .arrow-dropdown,
  :host(.ic-tree-item-disabled) ::slotted([slot="icon"]),
  :host(.ic-tree-item-disabled) ::slotted([slot="router-item"]) {
    color: GrayText !important;

    --ic-typography-color: GrayText !important;

    fill: GrayText !important;
  }

  :host .tree-item-content:focus,
  ::slotted([slot="router-item"]:focus) {
    border: var(--ic-hc-border) !important;
  }

  :host(.ic-tree-item-selected) .tree-item-content,
  :host(.ic-tree-item-selected) ::slotted([slot="router-item"]) {
    color: Highlight !important;

    --ic-typography-color: Highlight !important;
  }
}
