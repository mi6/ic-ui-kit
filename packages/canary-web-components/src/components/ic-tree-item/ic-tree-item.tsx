import {
  Component,
  h,
  Prop,
  Element,
  Event,
  EventEmitter,
  Host,
  Watch,
  State,
  Listen,
  forceUpdate,
  Method,
} from "@stencil/core";
import { IcSizes, IcThemeForegroundNoDefault } from "../../utils/types";
import {
  isSlotUsed,
  onComponentRequiredPropUndefined,
  checkSlotInChildMutations,
} from "../../utils/helpers";
import arrowDropdown from "../../assets/arrow-dropdown.svg";

let treeItemIds = 0;

/**
 * @slot label - Content is set as the tree item label.
 * @slot icon - Content is placed to the left of the label.
 * @slot router-item - Handle routing by nesting your routes in this slot.
 */
@Component({
  tag: "ic-tree-item",
  styleUrl: "ic-tree-item.css",
  shadow: true,
})
export class TreeItem {
  private treeItemId = `ic-tree-item-${treeItemIds++}`;
  private treeItemElement: HTMLElement;
  private treeItemTag = "IC-TREE-ITEM";
  private routerSlot: HTMLElement;
  private hostMutationObserver: MutationObserver = null;
  private TOOLTIP = "ic-tooltip";

  @Element() el: HTMLIcTreeItemElement;

  @State() childTreeItems: HTMLIcTreeItemElement[];

  /**
   * @internal Determines whether the light or dark variant of the tree item should be displayed.
   */
  @Prop() appearance?: IcThemeForegroundNoDefault = "dark";

  /**
   * If `true`, the tree item appears in the disabled state.
   */
  @Prop() disabled?: boolean = false;

  /**
   * @internal If `true`, the tree item appears in the expanded state.
   */
  @Prop({ mutable: true }) expanded: boolean = false;

  /**
   * @internal Determines if the parent tree item has been expanded.
   */
  @Prop({ mutable: true }) hasParentExpanded: boolean = false;

  /**
   * The URL that the tree item link points to. If set, the tree item will render as an "a" tag, otherwise it will render as a div.
   */
  @Prop() href?: string | undefined;

  /**
   * The human language of the linked URL.
   */
  @Prop() hreflang?: string = "";

  /**
   * @internal If `true`, the tree item is a parent of other tree items.
   */
  @Prop({ mutable: true }) isParent: boolean = false;

  /**
   * The label of the tree item.
   */
  @Prop() label: string = "";

  /**
   * How much of the referrer to send when following the link.
   */
  @Prop() referrerpolicy?: ReferrerPolicy;

  /**
   * The relationship of the linked URL as space-separated link types.
   */
  @Prop() rel?: string;

  /**
   * If `true`, the tree item appears in the selected state.
   */
  @Prop({ mutable: true }) selected: boolean = false;
  @Watch("selected")
  watchSelectedHandler() {
    if (this.selected) {
      this.icTreeItemSelected.emit({ id: this.el.id });
    }
    this.updateAriaLabel();
  }

  /**
   * @internal Determines the size of the tree item.
   */
  @Prop() size?: IcSizes = "medium";

  /**
   * The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe).
   */
  @Prop() target?: string;

  /**
   * Emitted when tree item is selected.
   */
  @Event() icTreeItemSelected: EventEmitter<{ id: string }>;

  disconnectedCallback(): void {
    this.hostMutationObserver?.disconnect();
  }

  componentDidLoad(): void {
    this.childTreeItems = Array.from((this.el as HTMLElement).children).filter(
      (child) => child.tagName === this.treeItemTag
    ) as HTMLIcTreeItemElement[];

    if (this.childTreeItems.length > 0) {
      this.isParent = true;
    }

    this.setTreeItemPadding();

    this.updateAriaLabel();

    this.truncateTreeItemLabel(this.el);

    !isSlotUsed(this.el, "label") &&
      onComponentRequiredPropUndefined(
        [{ prop: this.label, propName: "label" }],
        "Tree item"
      );

    this.hostMutationObserver = new MutationObserver(this.hostMutationCallback);
    this.hostMutationObserver.observe(this.el, {
      childList: true,
    });
  }

  componentDidUpdate(): void {
    if (this.hasParentExpanded) {
      this.childTreeItems.forEach((child) => {
        this.truncateTreeItemLabel(child);
      });
      this.hasParentExpanded = false;
    }
  }

  @Listen("keydown", {})
  handleKeyDown(ev: KeyboardEvent): void {
    if (ev.key === "Enter" || ev.key === " ") {
      ev.stopImmediatePropagation();
      this.handleTreeItemClicked();
    }
  }

  /**
   * Sets focus on the native `input`.
   */
  @Method()
  async setFocus(): Promise<void> {
    if (this.hasRouterSlot()) {
      this.routerSlot?.focus();
    } else {
      this.treeItemElement?.focus();
    }
  }

  private handleTreeItemClicked = (): void => {
    if (this.isParent) {
      this.expanded = !this.expanded;
      this.hasParentExpanded = true;
    }

    this.updateAriaLabel();
    this.selected = true;
    this.watchSelectedHandler();
  };

  /**
   * @internal Updates the aria-label of the tree item.
   */
  @Method()
  async updateAriaLabel(): Promise<void> {
    let ariaLabel;

    if (this.hasRouterSlot()) {
      ariaLabel = this.routerSlot.textContent;
    } else if (isSlotUsed(this.el, "label")) {
      ariaLabel = this.el.querySelector('[slot="label"]').textContent;
    } else {
      ariaLabel = this.label;
    }

    if (this.isParent) {
      ariaLabel = `${ariaLabel}, triggers submenu, ${
        this.expanded ? "expanded" : "collapsed"
      }`;
    }

    if (this.el.parentElement) {
      const treeItems = Array.from(
        (this.el.parentElement as HTMLElement).children
      ).filter(
        (child) => child.tagName === this.treeItemTag
      ) as HTMLIcTreeItemElement[];

      const index = treeItems.indexOf(this.el) + 1;
      const parentChildren = treeItems.length;

      ariaLabel = `${ariaLabel}, ${index} of ${parentChildren}`;
    }

    if (this.selected) {
      ariaLabel = `${ariaLabel}, active`;
    }

    if (this.disabled) {
      ariaLabel = `${ariaLabel}, dimmed`;
    }

    if (this.hasRouterSlot()) {
      this.routerSlot.ariaLabel = ariaLabel;
    } else {
      this.treeItemElement.ariaLabel = ariaLabel;
    }
  }

  private setTreeItemPadding = () => {
    let level = 1;
    let parentElement = this.el.parentElement;
    const treeItemContent = this.el.shadowRoot.querySelector(
      ".tree-item-content"
    ) as HTMLElement;

    const isSiblingOfParent = Array.from(parentElement.children)
      .map((sibling) => {
        if (
          sibling !== this.el &&
          !sibling.querySelector('[slot="router-item"]') &&
          !sibling.querySelector('[slot="label"]')
        ) {
          return sibling.children.length > 0;
        } else {
          return false;
        }
      })
      .includes(true);

    const isChild = parentElement.tagName === this.treeItemTag;

    const isRouterItem = this.hasRouterSlot();

    if (
      (isSiblingOfParent && !this.el.isParent) ||
      (isChild && !this.el.isParent)
    ) {
      if (isRouterItem) {
        this.routerSlot.classList.add("ic-tree-item-single");
      } else {
        treeItemContent.classList.add("ic-tree-item-single");
      }
    }

    while (parentElement) {
      if (parentElement.tagName === this.treeItemTag) {
        level++;
        if (!this.el.isParent && isSiblingOfParent) {
          treeItemContent.style.paddingLeft = `calc(var(--ic-space-xl) + ${
            level * 16
          }px)`;
        } else if (!this.el.isParent) {
          treeItemContent.style.paddingLeft = `calc(var(--ic-space-xs) + ${
            level * 24
          }px`;
        } else {
          treeItemContent.style.paddingLeft = `${level * 16}px`;
        }
      }
      parentElement = parentElement.parentElement;
    }
  };

  private truncateTreeItemLabel = (treeItem: HTMLIcTreeItemElement) => {
    const typographyEl: HTMLIcTypographyElement =
      treeItem.shadowRoot.querySelector(".tree-item-label");
    const tooltip = typographyEl?.closest(this.TOOLTIP);
    const treeContent: HTMLElement =
      treeItem.shadowRoot.querySelector(".tree-item-content");

    if (typographyEl?.scrollHeight > treeContent?.clientHeight) {
      typographyEl.classList.add("ic-text-overflow");

      if (!tooltip) {
        const tooltipEl = document.createElement("ic-tooltip");
        tooltipEl.setAttribute("target", this.el.id);
        tooltipEl.setAttribute("label", typographyEl.textContent);
        tooltipEl.classList.add("ic-tooltip-overflow");
        tooltipEl.setAttribute("placement", "right");
        treeContent.appendChild(tooltipEl);
        tooltipEl.appendChild(typographyEl);
      }
    }
  };

  private hasRouterSlot(): boolean {
    this.routerSlot = this.el.querySelector('[slot="router-item"]');
    return !!this.routerSlot;
  }

  private hostMutationCallback = (mutationList: MutationRecord[]): void => {
    if (
      mutationList.some(({ type, addedNodes, removedNodes }) =>
        type === "childList"
          ? checkSlotInChildMutations(addedNodes, removedNodes, "icon")
          : false
      )
    ) {
      forceUpdate(this);
    }
  };

  private handleDisplayTooltip = (display: boolean) => {
    const typographyEl: HTMLIcTypographyElement =
      this.el.shadowRoot.querySelector(".tree-item-label");
    const tooltip: HTMLIcTooltipElement = typographyEl?.closest(this.TOOLTIP);

    tooltip?.displayTooltip(display);
  };

  render() {
    const { appearance, disabled, label, selected, size, expanded } = this;

    const Component = this.href && !this.disabled ? "a" : "div";

    const attrs = Component == "a" && {
      href: this.href,
      hrefLang: this.hreflang,
      referrerPolicy: this.referrerpolicy,
      rel: this.rel,
      target: this.target,
    };

    return (
      <Host
        class={{
          [`ic-tree-item-${appearance}`]: true,
          "ic-tree-item-disabled": disabled,
          "ic-tree-item-selected": !disabled && selected,
          [`ic-tree-item-${size}`]: size !== "medium",
        }}
        id={this.treeItemId}
      >
        {this.hasRouterSlot() ? (
          <slot name="router-item" />
        ) : (
          <Component
            class={{
              "tree-item-content": true,
            }}
            tabIndex={disabled ? -1 : 0}
            onClick={this.handleTreeItemClicked}
            ref={(el) => (this.treeItemElement = el)}
            aria-disabled={disabled ? "true" : "false"}
            aria-live="polite"
            {...attrs}
            onFocus={() => this.handleDisplayTooltip(true)}
            onBlur={() => this.handleDisplayTooltip(false)}
          >
            {this.isParent && (
              <span
                class={{
                  ["arrow-dropdown"]: true,
                  ["tree-item-expanded"]: expanded,
                }}
                aria-hidden="true"
                innerHTML={arrowDropdown}
              />
            )}
            {isSlotUsed(this.el, "icon") && (
              <div class="icon-container">
                <slot name="icon" />
              </div>
            )}
            <ic-typography class="tree-item-label">
              {isSlotUsed(this.el, "label") ? <slot name="label" /> : label}
            </ic-typography>
          </Component>
        )}
        {expanded && (
          <div aria-hidden={`${!expanded}`}>
            <slot />
          </div>
        )}
      </Host>
    );
  }
}
