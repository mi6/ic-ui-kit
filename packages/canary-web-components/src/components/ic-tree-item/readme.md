# ic-tree-item



<!-- Auto Generated Below -->


## Properties

| Property         | Attribute        | Description                                                                                                                             | Type                                                                                                                                                                                     | Default     |
| ---------------- | ---------------- | --------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `disabled`       | `disabled`       | If `true`, the tree item appears in the disabled state.                                                                                 | `boolean`                                                                                                                                                                                | `false`     |
| `expanded`       | `expanded`       | If `true`, the tree item appears in the expanded state.                                                                                 | `boolean`                                                                                                                                                                                | `false`     |
| `href`           | `href`           | The URL that the tree item link points to. If set, the tree item will render as an "a" tag, otherwise it will render as a div.          | `string`                                                                                                                                                                                 | `undefined` |
| `hreflang`       | `hreflang`       | The human language of the linked URL.                                                                                                   | `string`                                                                                                                                                                                 | `""`        |
| `label`          | `label`          | The label of the tree item.                                                                                                             | `string`                                                                                                                                                                                 | `""`        |
| `referrerpolicy` | `referrerpolicy` | How much of the referrer to send when following the link.                                                                               | `"" \| "no-referrer" \| "no-referrer-when-downgrade" \| "origin" \| "origin-when-cross-origin" \| "same-origin" \| "strict-origin" \| "strict-origin-when-cross-origin" \| "unsafe-url"` | `undefined` |
| `rel`            | `rel`            | The relationship of the linked URL as space-separated link types.                                                                       | `string`                                                                                                                                                                                 | `undefined` |
| `selected`       | `selected`       | If `true`, the tree item appears in the selected state.                                                                                 | `boolean`                                                                                                                                                                                | `false`     |
| `target`         | `target`         | The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe).                                     | `string`                                                                                                                                                                                 | `undefined` |
| `theme`          | `theme`          | Sets the theme color to the dark or light theme color. "inherit" will set the color based on the system settings or ic-theme component. | `"dark" \| "inherit" \| "light"`                                                                                                                                                         | `"inherit"` |


## Events

| Event                | Description                         | Type                           |
| -------------------- | ----------------------------------- | ------------------------------ |
| `icTreeItemSelected` | Emitted when tree item is selected. | `CustomEvent<{ id: string; }>` |


## Methods

### `setFocus() => Promise<void>`

Sets focus on the native `input`.

#### Returns

Type: `Promise<void>`




## Slots

| Slot            | Description                                         |
| --------------- | --------------------------------------------------- |
| `"icon"`        | Content is placed to the left of the label.         |
| `"label"`       | Content is set as the tree item label.              |
| `"router-item"` | Handle routing by nesting your routes in this slot. |


## Dependencies

### Depends on

- ic-tooltip
- ic-typography

### Graph
```mermaid
graph TD;
  ic-tree-item --> ic-tooltip
  ic-tree-item --> ic-typography
  ic-tooltip --> ic-typography
  style ic-tree-item fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


