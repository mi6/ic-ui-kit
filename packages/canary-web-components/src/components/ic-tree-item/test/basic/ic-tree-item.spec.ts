import { newSpecPage } from "@stencil/core/testing";
import { TreeItem } from "../../ic-tree-item";
import { IcTypography as Typography } from "@ukic/web-components/dist/components/ic-typography";

describe("ic-tree-item component", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [TreeItem],
      html: `<ic-tree-item label="Item 1"></ic-tree-item>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with icon", async () => {
    const page = await newSpecPage({
      components: [TreeItem],
      html: `<ic-tree-item label="Item 1">
          <svg
            slot="icon"
            xmlns="http://www.w3.org/2000/svg"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
            fill="#000000"
          >
            <path d="M0 0h24v24H0V0z" fill="none" />
            <path
              d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
            />
          </svg>
        </ic-tree-item>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with child tree items", async () => {
    const page = await newSpecPage({
      components: [TreeItem],
      html: `<ic-tree-item label="Item 1">
          <ic-tree-item label="Item 1.1"></ic-tree-item>
          <ic-tree-item label="Item 1.2"></ic-tree-item>
        </ic-tree-item>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with router item slot", async () => {
    const page = await newSpecPage({
      components: [TreeItem],
      html: `<ic-tree-item label="Item 1">
          <a slot="router-item" href="/">Item 1</a>
        </ic-tree-item>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render disabled", async () => {
    const page = await newSpecPage({
      components: [TreeItem],
      html: `<ic-tree-item label="Item 1" disabled="true"></ic-tree-item>`,
    });

    expect(page.root).toMatchSnapshot();

    page.rootInstance.disabled = false;

    await page.waitForChanges();
    expect(page.root).toMatchSnapshot("disabled-removed");
  });

  it("should render selected", async () => {
    const page = await newSpecPage({
      components: [TreeItem],
      html: `<ic-tree-item label="Item 1" selected="true"></ic-tree-item>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with slotted label", async () => {
    const page = await newSpecPage({
      components: [TreeItem, Typography],
      html: `<ic-tree-item>
        <ic-typography slot="label" variant="body">Item 1</ic-typography>
      </ic-tree-item>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should select tree item on Enter", async () => {
    const page = await newSpecPage({
      components: [TreeItem],
      html: `<ic-tree-item label="Item 1"></ic-tree-item>`,
    });

    expect(page.rootInstance.selected).toBe(false);

    const event = new KeyboardEvent("keydown", { key: "Enter" });

    await page.rootInstance.handleKeyDown(event);

    expect(page.rootInstance.selected).toBe(true);
  });

  it("should expand/collapse parent tree item on Enter", async () => {
    const page = await newSpecPage({
      components: [TreeItem],
      html: `<ic-tree-item label="Item 1">
          <ic-tree-item label="Item 1.1"></ic-tree-item>
          <ic-tree-item label="Item 1.2"></ic-tree-item>
        </ic-tree-item>`,
    });

    expect(page.rootInstance.expanded).toBe(false);

    const event = new KeyboardEvent("keydown", { key: "Enter" });

    await page.rootInstance.handleKeyDown(event);

    expect(page.rootInstance.expanded).toBe(true);
  });

  it("should test rendering icon slot after initial render", async () => {
    const page = await newSpecPage({
      components: [TreeItem],
      html: `<ic-tree-item label="Item 1"></ic-tree-item>`,
    });

    const icon = document.createElement("svg");
    icon.setAttribute("slot", "icon");

    page.rootInstance.hostMutationCallback([
      {
        type: "childList",
        addedNodes: [icon],
        removedNodes: [],
      },
    ]);
  });
});
