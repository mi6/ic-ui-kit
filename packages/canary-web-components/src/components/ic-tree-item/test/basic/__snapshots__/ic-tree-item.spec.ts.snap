// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-tree-item component should render 1`] = `
<ic-tree-item aria-disabled="false" class="ic-tree-item-dark" id="ic-tree-item-0" label="Item 1">
  <mock:shadow-root>
    <div class="ic-tree-item-single tree-item-content" tabindex="0">
      <ic-typography class="tree-item-label">
        Item 1
      </ic-typography>
    </div>
  </mock:shadow-root>
</ic-tree-item>
`;

exports[`ic-tree-item component should render disabled 1`] = `
<ic-tree-item class="ic-tree-item-dark ic-tree-item-disabled" disabled="true" id="ic-tree-item-0" label="Item 1">
  <mock:shadow-root>
    <div aria-disabled="true" aria-live="polite" class="tree-item-content" tabindex="-1">
      <ic-typography class="tree-item-label">
        Item 1
      </ic-typography>
    </div>
  </mock:shadow-root>
</ic-tree-item>
`;

exports[`ic-tree-item component should render disabled: disabled-removed 1`] = `
<ic-tree-item class="ic-tree-item-dark" id="ic-tree-item-0" label="Item 1">
  <mock:shadow-root>
    <div aria-disabled="false" aria-live="polite" class="tree-item-content" tabindex="0">
      <ic-typography class="tree-item-label">
        Item 1
      </ic-typography>
    </div>
  </mock:shadow-root>
</ic-tree-item>
`;

exports[`ic-tree-item component should render selected 1`] = `
<ic-tree-item aria-disabled="false" class="ic-tree-item-dark ic-tree-item-selected" id="ic-tree-item-7" label="Item 1" selected="true">
  <mock:shadow-root>
    <div class="ic-tree-item-single tree-item-content" tabindex="0">
      <ic-typography class="tree-item-label">
        Item 1
      </ic-typography>
    </div>
  </mock:shadow-root>
</ic-tree-item>
`;

exports[`ic-tree-item component should render with child tree items 1`] = `
<ic-tree-item aria-disabled="false" class="ic-tree-item-dark" id="ic-tree-item-2" label="Item 1">
  <mock:shadow-root>
    <div class="tree-item-content" tabindex="0">
      <span aria-hidden="true" class="arrow-dropdown">
        svg
      </span>
      <ic-typography class="tree-item-label">
        Item 1
      </ic-typography>
    </div>
  </mock:shadow-root>
  <ic-tree-item aria-disabled="false" class="ic-tree-item-dark" id="ic-tree-item-3" label="Item 1.1">
    <mock:shadow-root>
      <div class="ic-tree-item-single tree-item-content" tabindex="0">
        <ic-typography class="tree-item-label">
          Item 1.1
        </ic-typography>
      </div>
    </mock:shadow-root>
  </ic-tree-item>
  <ic-tree-item aria-disabled="false" class="ic-tree-item-dark" id="ic-tree-item-4" label="Item 1.2">
    <mock:shadow-root>
      <div class="ic-tree-item-single tree-item-content" tabindex="0">
        <ic-typography class="tree-item-label">
          Item 1.2
        </ic-typography>
      </div>
    </mock:shadow-root>
  </ic-tree-item>
</ic-tree-item>
`;

exports[`ic-tree-item component should render with icon 1`] = `
<ic-tree-item aria-disabled="false" class="ic-tree-item-dark" id="ic-tree-item-1" label="Item 1">
  <mock:shadow-root>
    <div class="ic-tree-item-single tree-item-content" tabindex="0">
      <div class="icon-container">
        <slot name="icon"></slot>
      </div>
      <ic-typography class="tree-item-label">
        Item 1
      </ic-typography>
    </div>
  </mock:shadow-root>
  <svg fill="#000000" height="24px" slot="icon" viewBox="0 0 24 24" width="24px" xmlns="http://www.w3.org/2000/svg">
    <path d="M0 0h24v24H0V0z" fill="none"></path>
    <path d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"></path>
  </svg>
</ic-tree-item>
`;

exports[`ic-tree-item component should render with router item slot 1`] = `
<ic-tree-item aria-disabled="false" class="ic-tree-item-dark" id="ic-tree-item-5" label="Item 1">
  <mock:shadow-root>
    <slot name="router-item"></slot>
  </mock:shadow-root>
  <a href="/" slot="router-item">
    Item 1
  </a>
</ic-tree-item>
`;

exports[`ic-tree-item component should render with slotted label 1`] = `
<ic-tree-item aria-disabled="false" class="ic-tree-item-dark" id="ic-tree-item-8">
  <mock:shadow-root>
    <div class="ic-tree-item-single tree-item-content" tabindex="0">
      <ic-typography class="ic-typography-body tree-item-label">
        <mock:shadow-root>
          <slot></slot>
        </mock:shadow-root>
        <slot name="label"></slot>
      </ic-typography>
    </div>
  </mock:shadow-root>
  <ic-typography class="ic-typography-body" slot="label" variant="body">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    Item 1
  </ic-typography>
</ic-tree-item>
`;
